File: android/src/main/java/me/timeto/app/ui/home/tasks/HomeTaskView.kt
package me.timeto.app.ui.home.tasks

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.ExperimentalMaterialApi
import androidx.compose.material.Icon
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch
import me.timeto.app.Haptic
import me.timeto.app.R
import me.timeto.app.toColor
import me.timeto.app.ui.HStack
import me.timeto.app.ui.SwipeToAction
import me.timeto.app.ui.ZStack
import me.timeto.app.ui.c
import me.timeto.app.ui.home.HomeScreen__itemCircleFontSize
import me.timeto.app.ui.home.HomeScreen__itemCircleFontWeight
import me.timeto.app.ui.home.HomeScreen__itemCircleHPadding
import me.timeto.app.ui.home.HomeScreen__itemCircleHeight
import me.timeto.app.ui.home.HomeScreen__itemCircleMarginTrailing
import me.timeto.app.ui.home.HomeScreen__itemHeight
import me.timeto.app.ui.home.HomeScreen__primaryFontSize
import me.timeto.app.ui.navigation.LocalNavigationFs
import me.timeto.app.ui.onePx
import me.timeto.app.ui.roundedShape
import me.timeto.app.ui.symbol.SymbolView
import me.timeto.app.ui.task_timer.TaskTimerFs
import me.timeto.shared.ActivityUi
import me.timeto.shared.TextFeatures
import me.timeto.shared.db.TaskDb
import me.timeto.shared.vm.home.HomeVm
import me.timeto.shared.vm.home.tasks.HomeTasksItemUi

@OptIn(ExperimentalMaterialApi::class)
@Composable
fun HomeTaskView(
    homeTaskUi: HomeTasksItemUi.HomeTaskUi,
    homeState: HomeVm.State,
) {

    val navigationFs = LocalNavigationFs.current

    val isEditOrDelete = remember { mutableStateOf<Boolean?>(null) }
    val stateOffsetAbsDp = remember { mutableStateOf(0.dp) }

    val scope = rememberCoroutineScope()

    val taskDb: TaskDb =
        homeTaskUi.taskUi.taskDb

    SwipeToAction(
        isStartOrEnd = isEditOrDelete,
        modifier = Modifier
            .height(HomeScreen__itemHeight),
        startView = { state ->
            HomeTaskStaStartView(
                homeTaskUi = homeTaskUi,
                resetSta = { onComplete ->
                    scope.launch {
                        state.reset()
                        onComplete()
                    }
                },
            )
        },
        endView = { state ->
            HomeTaskStaEndView(
                state = state,
                onMoveToTimer = {
                    Haptic.long()
                    homeTaskUi.taskUi.moveToTimer()
                },
                onDelete = {
                    Haptic.long()
                    homeTaskUi.taskUi.delete()
                },
            )
        },
        onStart = {
            true
        },
        onEnd = {
            true
        },
        stateOffsetAbsDp = stateOffsetAbsDp,
        content = {
            HStack(
                modifier = Modifier
                    .height(HomeScreen__itemHeight)
                    .fillMaxWidth()
                    .background(c.black)
                    .padding(horizontal = homeTasksOuterHPadding)
                    .clip(roundedShape)
                    .clickable {
                        taskDb.startIntervalForUi(
                            ifJustStarted = {},
                            ifTimerNeeded = {
                                navigationFs.push {
                                    TaskTimerFs(
                                        taskDb = taskDb,
                                    )
                                }
                            },
                        )
                    }
                    .padding(start = homeTasksInnerHPadding),
                verticalAlignment = Alignment.CenterVertically,
            ) {

                val timeUi = homeTaskUi.timeUi
                if (timeUi != null) {
                    val bgColor: Color = when (timeUi.status) {
                        TextFeatures.TimeData.STATUS.IN -> c.homeFg
                        TextFeatures.TimeData.STATUS.SOON -> c.blue
                        TextFeatures.TimeData.STATUS.OVERDUE -> c.red
                    }
                    HStack(
                        modifier = Modifier
                            .padding(end = if (homeTaskUi.taskUi.tf.paused != null) 9.dp else HomeScreen__itemCircleMarginTrailing)
                            .height(HomeScreen__itemCircleHeight)
                            .clip(roundedShape)
                            .background(bgColor)
                            .padding(horizontal = HomeScreen__itemCircleHPadding),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(
                            text = timeUi.text,
                            modifier = Modifier
                                .padding(top = onePx),
                            fontWeight = HomeScreen__itemCircleFontWeight,
                            fontSize = HomeScreen__itemCircleFontSize,
                            lineHeight = 18.sp,
                            color = c.white,
                        )
                    }
                }

                if (homeTaskUi.taskUi.tf.paused != null) {
                    ZStack(
                        modifier = Modifier
                            .padding(end = 8.dp)
                            .size(HomeScreen__itemCircleHeight)
                            .clip(roundedShape)
                            .background(c.green),
                        contentAlignment = Alignment.Center,
                    ) {
                        Icon(
                            painter = painterResource(id = R.drawable.sf_pause_medium_black),
                            contentDescription = "Paused Task",
                            tint = c.white,
                            modifier = Modifier
                                .size(10.dp),
                        )
                    }
                }

                val activityUi: ActivityUi? =
                    homeTaskUi.taskUi.activityUi
                if (activityUi != null) {
                    ZStack(
                        modifier = Modifier
                            .width(HomeScreen__itemCircleHeight),
                    ) {
                        SymbolView(
                            symbol = activityUi.symbol,
                            color = remember(activityUi.colorRgba) {
                                activityUi.colorRgba.toColor()
                            },
                            letterSize = HomeScreen__primaryFontSize,
                            iconSize = 17.dp,
                            emojiSize = HomeScreen__itemCircleFontSize,
                            modifier = Modifier,
                        )
                    }
                }

                Text(
                    text = homeTaskUi.text,
                    modifier = Modifier
                        .padding(end = 4.dp)
                        .weight(1f),
                    fontSize = HomeScreen__primaryFontSize,
                    color = c.white,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )

                if (timeUi != null) {
                    val noteColor: Color = when (timeUi.status) {
                        TextFeatures.TimeData.STATUS.IN -> c.secondaryText
                        TextFeatures.TimeData.STATUS.SOON -> c.blue
                        TextFeatures.TimeData.STATUS.OVERDUE -> c.red
                    }
                    Text(
                        text = timeUi.note,
                        modifier = Modifier
                            .padding(end = homeTasksInnerHPadding),
                        fontSize = HomeScreen__primaryFontSize,
                        color = noteColor,
                    )
                }

                if ((homeState.taskFolderUi.activityDb != null) &&
                    (taskDb.isToday || taskDb.isTomorrow)
                ) {
                    HomeTasksFolderButton(
                        taskFolderUi = homeTaskUi.taskUi.taskFolderUi,
                        color = if (taskDb.isToday) c.orange else c.indigo,
                        modifier = Modifier
                            .padding(end = 1.dp),
                        onClick = {
                            homeTaskUi.taskUi.updateTaskFolder(
                                taskFolderDb = homeState.taskFolderUi.taskFolderDb,
                            )
                        },
                    )
                }
            }
        },
    )
}

File: apple/iphone/UI/Home/Tasks/HomeTaskView.swift
import SwiftUI
import shared

struct HomeTaskView: View {
    
    let homeTaskUi: HomeTasksItemUi.HomeTaskUi
    let homeModeTaskFolder: HomeMode.TaskFolder
    
    ///
    
    @Environment(Navigation.self) private var navigation
    
    @State private var xSwipeOffset: CGFloat = 0
    @State private var width: CGFloat? = nil
    
    var body: some View {
        
        let taskDb: TaskDb = homeTaskUi.taskUi.taskDb
        
        ZStack {
            
            GeometryReader { proxy in
                ZStack {
                }
                .onChange(of: proxy.size.width, initial: true) { _, newValue in
                    width = newValue
                }
                .onChange(of: proxy.frame(in: .global).minY) {
                    xSwipeOffset = 0
                }
            }
            .frame(height: 1) // height: 1, or full height items if short
            
            if (xSwipeOffset < 0) {
                HomeTaskStaEndView(
                    onMoveToTimer: {
                        Haptic.mediumShot()
                        homeTaskUi.taskUi.moveToTimer()
                    },
                    onDelete: {
                        Haptic.mediumShot()
                        homeTaskUi.taskUi.delete()
                    },
                    onCancel: {
                        withAnimation {
                            xSwipeOffset = 0
                        }
                    },
                )
            } else if (xSwipeOffset > 0) {
                HomeTaskStaStartView(
                    homeTaskUi: homeTaskUi,
                    onCancel: {
                        withAnimation {
                            xSwipeOffset = 0
                        }
                    }
                )
            }
            
            ///

            Button(
                action: {
                    taskDb.startIntervalForUi(
                        ifJustStarted: {},
                        ifTimerNeeded: {
                            navigation.showTaskTimerSheet(
                                taskDb: taskDb,
                            )
                        },
                    )
                },
                label: {
                    
                    HStack {
                        
                        if let timeUi = homeTaskUi.timeUi {
                            let bgColor: Color = switch timeUi.status {
                            case .in: homeFgColor
                            case .soon: .blue
                            case .overdue: .red
                            default: fatalError("timeUi.status bgColor not handled")
                            }
                            Text(timeUi.text)
                                .foregroundColor(.white)
                                .font(.system(size: HomeScreen__itemCircleFontSize, weight: HomeScreen__itemCircleFontWeight))
                                .padding(.horizontal, HomeScreen__itemCircleHPadding)
                                .frame(height: HomeScreen__itemCircleHeight)
                                .background(roundedShape.fill(bgColor))
                                .padding(.trailing, homeTaskUi.taskUi.tf.paused != nil ? 9 : HomeScreen__itemCircleMarginTrailing)
                        }
                        
                        if homeTaskUi.taskUi.tf.paused != nil {
                            ZStack {
                                Image(systemName: "pause")
                                    .foregroundColor(.white)
                                    .font(.system(size: 12, weight: .black))
                            }
                            .frame(width: HomeScreen__itemCircleHeight, height: HomeScreen__itemCircleHeight)
                            .background(roundedShape.fill(.green))
                            .padding(.trailing, 8)
                        }
                        
                        if let activityUi = homeTaskUi.taskUi.activityUi {
                            let symbol: Symbol = activityUi.symbol
                            HStack {
                                
                                let offsetX: CGFloat = {
                                    if symbol is Symbol.Letter {
                                        return 0
                                    }
                                    if symbol is Symbol.Icon {
                                        return -2
                                    }
                                    if symbol is Symbol.Emoji {
                                        return -2
                                    }
                                    return 0
                                }()
                                
                                SymbolView(
                                    symbol: symbol,
                                    color: activityUi.colorRgba.toColor(),
                                    letterSize: HomeScreen__primaryFontSize,
                                    iconSize: 15,
                                    emojiSize: HomeScreen__itemCircleFontSize,
                                )
                                .offset(x: offsetX)
                                
                                Spacer()
                            }
                            .frame(width: 20)
                        }
                        
                        Text(homeTaskUi.text)
                            .font(.system(size: HomeScreen__primaryFontSize))
                            .foregroundColor(.white)
                            .padding(.trailing, 4)
                        
                        Spacer()
                        
                        if let timeUi = homeTaskUi.timeUi {
                            let noteColor: Color = switch timeUi.status {
                            case .in: .secondary
                            case .soon: .blue
                            case .overdue: .red
                            default: fatalError("timeUi.status noteColor not handled")
                            }
                            Text(timeUi.note)
                                .foregroundColor(noteColor)
                                .font(.system(size: HomeScreen__primaryFontSize))
                                .padding(.trailing, HomeScreen__itemCircleHPadding)
                        }
                        
                        if (homeModeTaskFolder.taskFolderDb.activity_id != nil) && (taskDb.isToday || taskDb.isTomorrow) {
                            HomeBarTaskFolderButton(
                                taskFolderUi: homeTaskUi.taskUi.taskFolderUi,
                                color: taskDb.isToday ? .orange : .indigo,
                                onClick: {
                                    homeTaskUi.taskUi.updateTaskFolder(
                                        taskFolderDb: homeModeTaskFolder.taskFolderDb,
                                    )
                                },
                            )
                        }
                    }
                    .frame(height: HomeScreen__itemHeight)
                    .padding(.leading, HomeScreen__hPadding)
                }
            )
            .background(.black)
            .offset(x: xSwipeOffset)
            .highPriorityGesture(gesture)
        }
    }
    
    // https://stackoverflow.com/a/79037514
    private var gesture: some Gesture {
        DragGesture(minimumDistance: 25, coordinateSpace: .global)
            .onChanged { value in
                xSwipeOffset = value.translation.width
            }
            .onEnded { value in
                if value.translation.width < -80 {
                    xSwipeOffset = (width ?? 999) * -1
                } else if value.translation.width > 60 {
                    xSwipeOffset = width ?? 999
                } else {
                    xSwipeOffset = 0
                }
            }
    }
}
